hyp3_gamma/dem.py

```diff
@@ -0,0 +1,130 @@
+import json
+from pathlib import Path
+from subprocess import PIPE, run
+from tempfile import TemporaryDirectory
+from typing import Generator, List, Tuple
+
+from hyp3lib import DemError
+from hyp3lib.execute import execute
+from osgeo import gdal, ogr
+
+DEM_GEOJSON = '/vsicurl/https://asf-dem-west.s3.us-west-2.amazonaws.com/v2/cop30.geojson'
+
+gdal.UseExceptions()
+ogr.UseExceptions()
+
+
+def get_geometry_from_kml(kml_file: str) -> ogr.Geometry:
+    response = run(['ogr2ogr', '-wrapdateline', '-datelineoffset', '20', '-f', 'GeoJSON', '-mapfieldtype',
+                    'DateTime=String', '/vsistdout', kml_file], stdout=PIPE, check=True)
+    geojson = json.loads(response.stdout)
+    geometry = json.dumps(geojson['features'][0]['geometry'])
+    return ogr.CreateGeometryFromJson(geometry)
+
+
+def get_dem_features() -> Generator[ogr.Feature, None, None]:
+    ds = ogr.Open(DEM_GEOJSON)
```

Comment:
this might not get properly closed, if `ds` has a method `ds.close` this can be resolved with
```py
from contextlib import closing
def get_dem_features() -> Generator[ogr.Feature, None, None]:
    ds = ogr.Open(DEM_GEOJSON)
    with closing(ds):
        ....
```
alternatively, we can write out own context manager
```py
@context_manager
def ogr_file_managed(file_path):
    ds = ogr.Open(file_path)
    try:
        yield ds
    finally:
        del ds

def get_dem_features() -> Generator[ogr.Feature, None, None]:
    with ogr_file_managed(DEM_GEOJSON) as ds:
        ....
```

Reply:
Nevermind, `ds` will be cleaned up when teh function that calls it goes out of scope

Reply:
GDAL shit is weird; you have to dereference to close. 

```python
from osgeo import ogr
DEM_GEOJSON = '/vsicurl/https://asf-dem-west.s3.us-west-2.amazonaws.com/v2/cop30.geojson'
ds = ogr.Open(DEM_GEOJSON)
ds.close()
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "/home/jhkennedy/mambaforge/envs/hyp3-autorift/lib/python3.8/site-packages/osgeo/ogr.py", line 452, in <lambda>
    __getattr__ = lambda self, name: _swig_getattr(self, DataSource, name)
  File "/home/jhkennedy/mambaforge/envs/hyp3-autorift/lib/python3.8/site-packages/osgeo/ogr.py", line 80, in _swig_getattr
    raise AttributeError("'%s' object has no attribute '%s'" % (class_type.__name__, name))
AttributeError: 'DataSource' object has no attribute 'close'
```

You can implement a context manager, and that's what [rasterio does](https://github.com/mapbox/rasterio/blob/master/rasterio/__init__.py#L43), but [it's complicated](https://stackoverflow.com/questions/49648360/use-with-statement-to-close-gdal-datasets) to do well for all GDAL things. 

If we want things like context mangers and what not, we should use [rasterio  instead](https://rasterio.readthedocs.io/en/latest/topics/switch.html) of vanilla gdal, but [it's all in or nothing](https://rasterio.readthedocs.io/en/latest/topics/switch.html#mutual-incompatibilities:~:text=Rasterio%20and%20GDAL%E2%80%99s%20bindings%20can%20contend,of%20import%20osgeo.gdal%20or%20import%20rasterio.) and we do enough with GDAL that switching would be a big lift.